File: Python3/0712_minimum_ascii_delete_sum_for_two_strings.py
# https://leetcode.cn/problems/minimum-ascii-delete-sum-for-two-strings/

class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        n1, n2 = len(s1), len(s2)
        dp = [[0 for _ in range(n2 + 1)] for _ in range(n1 + 1)]
        for i1 in range(1, n1+1):
            dp[i1][0] = dp[i1-1][0] + ord(s1[i1-1])
        for i2 in range(1, n2+1):
            dp[0][i2] = dp[0][i2-1] + ord(s2[i2-1])
        for i1 in range(1, n1+1):
            for i2 in range(1, n2+1):
                if s1[i1-1] == s2[i2-1]:
                    dp[i1][i2] = dp[i1-1][i2-1]
                else:
                    dp[i1][i2] = min(dp[i1-1][i2] + ord(s1[i1-1]),
                                     dp[i1][i2-1] + ord(s2[i2-1]))
        return dp[n1][n2]

File: Cpp/0712_minimum_ascii_delete_sum_for_two_strings.cpp
// https://leetcode.cn/problems/minimum-ascii-delete-sum-for-two-strings/

#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    int minimumDeleteSum(string s1, string s2) {
        int n1 = s1.size(), n2 = s2.size();
        vector<vector<int>> dp(n1+1, vector<int>(n2+1, 0));
        for (int i1 = 1; i1 <= n1; ++i1) {
            dp[i1][0] = dp[i1-1][0] + static_cast<int>(s1[i1-1]);
        }
        for (int i2 = 1; i2 <= n2; ++i2) {
            dp[0][i2] = dp[0][i2-1] + static_cast<int>(s2[i2-1]);
        }
        for (int i1 = 1; i1 <= n1; ++i1) {
            for (int i2 = 1; i2 <= n2; ++i2) {
                if (s1[i1-1] == s2[i2-1]) {
                    dp[i1][i2] = dp[i1-1][i2-1];
                } else {
                    dp[i1][i2] = min(dp[i1-1][i2] + static_cast<int>(s1[i1-1]),
                                     dp[i1][i2-1] + static_cast<int>(s2[i2-1]));
                }
            }
        }
        return dp[n1][n2];
    }
};

int main() {
    Solution sol;
    string s1 = "sea";
    string s2 = "eat";
    cout << sol.minimumDeleteSum(s1, s2) << '\n'; // Output: 231
    return 0;
}